File: ios/Tests/BitchatMeshTests/GeohashTests.swift
import XCTest
@testable import BitchatMesh

final class GeohashTests: XCTestCase {
    func testGeohashEncodeDecodeBoundsContainPoint() {
        let lat = 37.7749
        let lon = -122.4194
        let precision = 8

        let geohash = Geohash.encode(latitude: lat, longitude: lon, precision: precision)
        XCTAssertEqual(geohash.count, precision)
        XCTAssertTrue(Geohash.isValidBuildingGeohash(geohash))

        let bounds = Geohash.decodeBounds(geohash)
        XCTAssertGreaterThanOrEqual(lat, bounds.latMin)
        XCTAssertLessThanOrEqual(lat, bounds.latMax)
        XCTAssertGreaterThanOrEqual(lon, bounds.lonMin)
        XCTAssertLessThanOrEqual(lon, bounds.lonMax)
    }

    func testGeohashNeighborsAreValidAndDistinct() {
        let geohash = Geohash.encode(latitude: 37.7749, longitude: -122.4194, precision: 8)
        let neighbors = Geohash.neighbors(of: geohash)

        XCTAssertEqual(neighbors.count, 8)
        XCTAssertFalse(neighbors.contains(geohash))
        XCTAssertEqual(Set(neighbors).count, 8)
        XCTAssertTrue(neighbors.allSatisfy { $0.count == 8 })
        XCTAssertTrue(neighbors.allSatisfy { Geohash.isValidBuildingGeohash($0) })
    }

    func testGeohashValidationRejectsInvalidLengthAndCharacters() {
        XCTAssertFalse(Geohash.isValidBuildingGeohash(""))
        XCTAssertFalse(Geohash.isValidBuildingGeohash("abc"))
        XCTAssertFalse(Geohash.isValidBuildingGeohash("abc123"))

        XCTAssertFalse(Geohash.isValidBuildingGeohash("abcd123!"))
        XCTAssertFalse(Geohash.isValidBuildingGeohash("abcd123o"))
        XCTAssertFalse(Geohash.isValidBuildingGeohash("abcd123i"))
    }

    func testGeohashEncodeReturnsEmptyWhenPrecisionIsNonPositive() {
        XCTAssertEqual(Geohash.encode(latitude: 0.0, longitude: 0.0, precision: 0), "")
        XCTAssertEqual(Geohash.encode(latitude: 0.0, longitude: 0.0, precision: -3), "")
    }

    func testGeohashDecodeCenterReencodesToSameCell() {
        let precision = 8
        let geohash = Geohash.encode(latitude: 37.7749, longitude: -122.4194, precision: precision)
        let center = Geohash.decodeCenter(geohash)
        let reencoded = Geohash.encode(latitude: center.lat, longitude: center.lon, precision: precision)

        XCTAssertEqual(reencoded, geohash)
    }

    func testGeohashNeighborsNearPoleSkipOutOfBoundsCells() {
        let geohash = Geohash.encode(latitude: 89.9999, longitude: 0.0, precision: 8)
        let neighbors = Geohash.neighbors(of: geohash)

        XCTAssertGreaterThan(neighbors.count, 0)
        XCTAssertLessThan(neighbors.count, 8)
        XCTAssertEqual(Set(neighbors).count, neighbors.count)
        XCTAssertTrue(neighbors.allSatisfy { $0.count == 8 })
        XCTAssertTrue(neighbors.allSatisfy { Geohash.isValidBuildingGeohash($0) })
    }
}

File: android/bitchat-mesh/src/test/kotlin/com/bitchat/android/protocol/GeohashTests.kt
package com.bitchat.android.protocol

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class GeohashTests {
    @Test
    fun testGeohashEncodeDecodeBoundsContainPoint() {
        val lat = 37.7749
        val lon = -122.4194
        val precision = 8

        val geohash = Geohash.encode(latitude = lat, longitude = lon, precision = precision)
        assertEquals(precision, geohash.length)
        assertTrue(Geohash.isValidBuildingGeohash(geohash))

        val bounds = Geohash.decodeBounds(geohash)
        assertTrue(lat >= bounds.latMin)
        assertTrue(lat <= bounds.latMax)
        assertTrue(lon >= bounds.lonMin)
        assertTrue(lon <= bounds.lonMax)
    }

    @Test
    fun testGeohashNeighborsAreValidAndDistinct() {
        val geohash = Geohash.encode(latitude = 37.7749, longitude = -122.4194, precision = 8)
        val neighbors = Geohash.neighbors(geohash)

        assertEquals(8, neighbors.size)
        assertFalse(neighbors.contains(geohash))
        assertEquals(8, neighbors.toSet().size)
        assertTrue(neighbors.all { it.length == 8 })
        assertTrue(neighbors.all { Geohash.isValidBuildingGeohash(it) })
    }

    @Test
    fun testGeohashValidationRejectsInvalidLengthAndCharacters() {
        assertFalse(Geohash.isValidBuildingGeohash(""))
        assertFalse(Geohash.isValidBuildingGeohash("abc"))
        assertFalse(Geohash.isValidBuildingGeohash("abc123"))
        assertFalse(Geohash.isValidBuildingGeohash("abcd123!"))
        assertFalse(Geohash.isValidBuildingGeohash("abcd123o"))
        assertFalse(Geohash.isValidBuildingGeohash("abcd123i"))
    }

    @Test
    fun testGeohashEncodeReturnsEmptyWhenPrecisionIsNonPositive() {
        assertEquals("", Geohash.encode(latitude = 0.0, longitude = 0.0, precision = 0))
        assertEquals("", Geohash.encode(latitude = 0.0, longitude = 0.0, precision = -3))
    }

    @Test
    fun testGeohashDecodeCenterReencodesToSameCell() {
        val precision = 8
        val geohash = Geohash.encode(latitude = 37.7749, longitude = -122.4194, precision = precision)
        val center = Geohash.decodeCenter(geohash)
        val reencoded = Geohash.encode(latitude = center.lat, longitude = center.lon, precision = precision)

        assertEquals(geohash, reencoded)
    }

    @Test
    fun testGeohashNeighborsNearPoleSkipOutOfBoundsCells() {
        val geohash = Geohash.encode(latitude = 89.9999, longitude = 0.0, precision = 8)
        val neighbors = Geohash.neighbors(geohash)

        assertTrue(neighbors.isNotEmpty())
        assertTrue(neighbors.size < 8)
        assertEquals(neighbors.toSet().size, neighbors.size)
        assertTrue(neighbors.all { it.length == 8 })
        assertTrue(neighbors.all { Geohash.isValidBuildingGeohash(it) })
    }
}
